### ISSUE TITLE
SH4: MMU Support

### ISSUE THREAD

**Issue by MrPsyMan** [reicast/reicast-emulator#53]
*edited by @skmp*
SH4 has support for full memory remapping, and this is currently not implemented in reicast. nullDC included some parts of it, but the support was mostly buggy and was never supported for the JIT. From my knowledge, only Makaron handles this.

Windows CE based games, Bleemcast, Linux for Dreamcast and NetBSD for Dreamcast use the mmu to its full potential. Examples: Virtua Cop 2. Complete list at https://segaretro.org/Windows_CE

Some Katana-SDK games use a subset of SQ remapping, which is implemented. Examples: Ikaruga

I imported most of the mmu code from nullDC to this project in #818 and d99c803a0cd8f5b8dcc5a9b9fce5055b7c521cba, however, I have really tested it.

Getting mmu working is a /very/ complicated task, and integrating it in the JITs will take major work the ugliest & most complicated areas of reicast.

We need a brave contributor to look further into this :)

---

Again some wishful thinking.

MMU is not supported currently. This has to change.

**Comment by hooby3dfx** [comment]
was just doing some research and as mentioned above by @MrPsyMan there seems to be varying levels of support for mmu, with simpler remapping correcting the 2k sports games and ooga booga.
@skmp looks like this was done in nulldc here: 
https://code.google.com/p/nulldc/issues/detail?id=66&can=1&q=nba&colspec=ID%20Type%20Status%20Priority%20Milestone%20Owner%20Summary%20emu
https://code.google.com/p/nulldc/source/detail?r=140

any chance of this getting corrected in reicast?

**Comment by skmp** [comment]
I recall that, it was a fascinating bug fix. *checks if it is included in reicast*

**Comment by skmp** [comment]
Yes, the arm bug fix is included in reicast. I'm not particularly sure how it affects the arm7 jit, if at all. MMU support is a major chunk of work that is likely to be very slow, so it's still in the "would be nice to have" plans

**Comment by hooby3dfx** [comment]
Those games still appeared to have that issue last I checked. Will retest

**Comment by skmp** [comment]
This has been partially implemented for the interpreter by re-using nullDC's mmu implementation. Not tested though.

**Comment by MrPsyMan** [comment]
We want a framework setup for this and leave implementation for later.

Setting up as mid term goal for now.

**Comment by skmp** [comment]
Mostly implemented after #1523
